### PR TITLE
Make hard memory limit configurable

### LINKED ISSUE #509: Extend instantiate() with a memory limit
This would replace the hard `MemoryPagesLimit` in `limits.h`. If no limit is set by the user, it defaults to `MemoryPagesLimit` (256Mb). If it is set higher than 2^32-1 return an error (or just make the limit uint32).

Also set this limit to 4Gb in spectests in order to pass them.

### PR DESCRIPTION
**Pull request by gumb0**
Closes #509

### CODE REVIEW

File: lib/fizzy/execute.cpp [hunk 1]
```diff
@@ -1229,9 +1230,10 @@ ExecutionResult execute(
             uint32_t ret = static_cast<uint32_t>(cur_pages);
             try
             {
+                // TODO use memory_pages_limit passed to instantiate
                 const size_t memory_max_pages =
```
Comment (axic) [inline lib/fizzy/execute.cpp:1234]: Couldn't we update `instance.memory_pages_limit` with the imported max?

File: test/unittests/instantiate_test.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ TEST(instantiate, imported_memory_invalid)
     const auto bin_without_max = from_hex("0061736d01000000020a01036d6f64016d020001");
     const auto module_without_max = parse(bin_without_max);
     EXPECT_THROW_MESSAGE(
-        instantiate(module_without_max, {}, {}, {{&memory, {1, MemoryPagesLimit + 1}}}),
+        instantiate(module_without_max, {}, {}, {{&memory, {1, DefaultMemoryPagesLimit + 1}}}),
         instantiate_error,
         "imported memory limits cannot exceed hard memory limit of 268435456 bytes");
```
Comment (axic) [inline test/unittests/instantiate_test.cpp:302]: Text update?

**Comment by axic** [comment]
Looks good in general, this is what I meant.

File: lib/fizzy/execute.cpp (hunk 1 above)
Comment (gumb0) [reply lib/fizzy/execute.cpp:1234]: I'll try it

File: lib/fizzy/execute.cpp [hunk 3]
```diff
@@ -701,7 +701,13 @@ std::unique_ptr<Instance> instantiate(Module module,
 
     auto [table, table_limits] = allocate_table(module.tablesec, imported_tables);
 
-    auto [memory, memory_limits] = allocate_memory(module.memorysec, imported_memories);
+    auto [memory, memory_limits] =
+        allocate_memory(module.memorysec, imported_memories, memory_pages_limit);
+    if (memory_limits.max.has_value())
```
Comment (axic) [inline lib/fizzy/execute.cpp:706]: Would add something like:
```suggestion
    // If imported memory is used, we need to adjust the memory limit.
    // Note: allocate_memory ensures this limit is always below memory_pages_limit.
    if (memory_limits.max.has_value())
```

File: test/spectests/spectests.cpp [hunk 4]
```diff
@@ -126,9 +127,10 @@ class test_runner
                         continue;
                     }
 
-                    m_instances[name] = fizzy::instantiate(std::move(module),
-                        std::move(imports.functions), std::move(imports.tables),
-                        std::move(imports.memories), std::move(imports.globals));
+                    m_instances[name] =
+                        fizzy::instantiate(std::move(module), std::move(imports.functions),
+                            std::move(imports.tables), std::move(imports.memories),
+                            std::move(imports.globals), TestMemoryPagesLimit);
```
Comment (axic) [inline test/spectests/spectests.cpp:133]: I'll actually override the default in #329 too.

File: lib/fizzy/execute.cpp (hunk 3 above)
Comment (gumb0) [reply lib/fizzy/execute.cpp:709]: I added the comment, but it applies to non-imported memory, too.

Comment (axic) [reply lib/fizzy/execute.cpp:709]: Ah right, it is the memory section setting.

File: lib/fizzy/execute.cpp [hunk 5]
```diff
@@ -173,7 +172,7 @@ std::tuple<table_ptr, Limits> allocate_table(
 }
 
 std::tuple<bytes_ptr, Limits> allocate_memory(const std::vector<Memory>& module_memories,
-    const std::vector<ExternalMemory>& imported_memories)
+    const std::vector<ExternalMemory>& imported_memories, unsigned memory_pages_limit)
```
Comment (chfast) [inline lib/fizzy/execute.cpp:175]:
```suggestion
    const std::vector<ExternalMemory>& imported_memories, uint32_t memory_pages_limit)
```
Always use uint32_t for memory indexing and sizes?

File: lib/fizzy/execute.cpp [hunk 6]
```diff
@@ -677,7 +676,8 @@ std::optional<uint32_t> find_export(const Module& module, ExternalKind kind, std
 
 std::unique_ptr<Instance> instantiate(Module module,
     std::vector<ExternalFunction> imported_functions, std::vector<ExternalTable> imported_tables,
-    std::vector<ExternalMemory> imported_memories, std::vector<ExternalGlobal> imported_globals)
+    std::vector<ExternalMemory> imported_memories, std::vector<ExternalGlobal> imported_globals,
+    unsigned memory_pages_limit /* = DefaultMemoryPagesLimit */)
```
Comment (chfast) [inline lib/fizzy/execute.cpp:680]:
```suggestion
    unsigned memory_pages_limit /*= DefaultMemoryPagesLimit*/)
```
To be consistent with `/*unused*/`?

File: lib/fizzy/execute.cpp (hunk 5 above)
Comment (axic) [reply lib/fizzy/execute.cpp:175]: This is number of pages, the limits sections uses `uint32_t`, and we actually use `size_t` in multiple places regarding this.

**Review by chfast — APPROVED** [review]

**Review by axic — APPROVED** [review]

**Pull request merged**